xla/client/lib/math.cc

```diff
@@ -1175,10 +1175,130 @@ XlaOp Acos(XlaOp x) {
+      //
+      // Hull et al provide an algorithm for computing complex arcsin
+      // that is based on splitting the first quarter of the complex
+      // plane into 11 regions and providing its own approximations
+      // to each region. The evaluation of arcsin in other quarters
+      // is resolved by using identities
+      //   arcsin(-z) == -arcsin(z)
+      //   arcsin(conj(z)) == conj(arcsin(z))
+      //
+      // When considering the evaluation of arcsin real and imaginary
+      // parts separately, the 11 regions by Hull et al can be reduced
+      // to 3 regions for the real part and 6 regions for the
+      // imaginary part. In the following, the corresponding
+      // modification of the Hull et al algorithm is provided. It is
+      // validated against the mpmath.mp.asin that implements the
+      // original Hull et al algorithm.
+
+      auto dtype = primitive_util::ComplexComponentType(shape.element_type());
+
+      double safe_min_, safe_max_;
+      switch (dtype) {
+        case F64:
+          safe_min_ = std::sqrt(std::numeric_limits<double>::min()) * 4;
+          safe_max_ = std::sqrt(std::numeric_limits<double>::max()) / 8;
+          break;
+        case F32:
+          safe_min_ = std::sqrt(std::numeric_limits<float>::min()) * 4;
+          safe_max_ = std::sqrt(std::numeric_limits<float>::max()) / 8;
+          break;
+        default:
+          safe_min_ = safe_max_ = 0.0;
+          assert(false);  // unreachable
+      }
+      auto signed_x = Real(z);
+      auto signed_y = Imag(z);
+      auto x = Abs(signed_x);
+      auto y = Abs(signed_y);
+
+      auto safe_min = ScalarLike(x, safe_min_);
+      auto safe_max = ScalarLike(x, safe_max_);
+      auto safe_max_m6 = ScalarLike(x, safe_max_ * 1e-6);
+      auto safe_max_p12 = ScalarLike(x, safe_max_ * 1e12);
+      auto safe_max_p2 = ScalarLike(x, safe_max_ * 1e2);
+      auto safe_max_opt = Select(Lt(x, safe_max_p12), safe_max_m6, safe_max_p2);
+      auto y_ge_max_opt = Ge(y, safe_max_opt);
+
+      auto zero = ScalarLike(x, 0);
+      auto one = ScalarLike(x, 1);
+      auto none = ScalarLike(x, -1);
+      auto half = ScalarLike(x, 0.5);
+      auto log2 = ScalarLike(x, std::log(2.0));
+      auto xp1 = Add(x, one);
+      auto xm1 = Add(x, none);
+      auto yy = Mul(y, y);
+      auto half_yy = Mul(half, yy);
+      auto r = Hypot(xp1, y);
+      auto s = Hypot(xm1, y);
+      auto a = Mul(half, Add(r, s));
+      auto ap1 = Add(a, one);
+      auto half_apx = Mul(half, Add(a, x));
+      auto rpxp1 = Add(r, xp1);
+      auto spxm1 = Add(s, xm1);
+      auto smxm1 = Sub(s, xm1);
+      auto y_is_finite = Not(IsPosInf(y));
+      auto max_xy = Select(Gt(x, y), x, y);
+
+      auto y1 = Select(
+          Gt(max_xy, safe_max), y,
+          Select(
+              Le(x, one), Sqrt(Mul(half_apx, Add(Div(yy, rpxp1), smxm1))),
+              Mul(y, Sqrt(Add(Div(half_apx, rpxp1), Div(half_apx, spxm1))))));
+      auto real = Atan2(signed_x, y1);
```

Comment:
I'm having trouble understanding this in terms of the formula in the comment above – I would have expected the real part of the return value to be `arcsin(x / a)`. Is the reasoning here that `y1` is constructed so that `x ** 2 + y1 ** 2 == a ** 2`, and therefore `arcsin(x / a)` is equivalent to `arctan(x/y1)`? If so, it may warrant a comment.

Reply:
First, the motivation for formulating the approximations from different regions in terms of `arctan2` is to reduce the number of calls to math functions to a minimal.
In the Hull et al paper, the real part is evaluated using a number of approximations that either call `arcsin` or `arctan` or simply `x / y`. It turns out that in specific regions, expressions with `arcsin` or division can be expressed in terms of `arctan2` and in other regions the approximations already use `arctan2`. Also, some regions using `arctan2` can be extended to cover certain parts of regions using `arcsin` without loosing accuracy for the given floating point system (float64, float32, ...).

While the Hull et al paper provides the mathematical proof to the implemented algorithm, there exists conditions where Hull et al are too conservative when considering that we are dealing with finite-precision numbers. MPMath implements the original Hull et al algorithm with multi-precision arithmetic that here we are using as reference for validating the modified Hull et al algorithm.

I'll add more comments explaining what is going on.

Reply:
See asin algorithm description in math_impl.h, in particular the notes the real part.